### PR TITLE
use fork id to filter out non matching peers

### PR DESCRIPTION
**Pull request by pinges**
After bonding with a new peer request ENR, and if the ENR response contains the ForkId, only try to connect if the ForkId matches. Besu will still try to connect to peers not sending the ForkId.

Signed-off-by: Stefan <stefan.pingel@consensys.net>

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java [hunk 1]
```diff
@@ -1176,4 +1176,9 @@ public RunnerBuilder minPeers(final int minPeers) {
     this.minPeers = minPeers;
     return this;
   }
+
+  public RunnerBuilder legacyForkId(final boolean legacyEth64ForkIdEnabled) {
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java:1180]: can you explain legacy fork Id?

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java [hunk 2]
```diff
@@ -144,7 +146,7 @@ public class DefaultP2PNetwork implements P2PNetwork {
   private final AtomicBoolean started = new AtomicBoolean(false);
   private final AtomicBoolean stopped = new AtomicBoolean(false);
   private final CountDownLatch shutdownLatch = new CountDownLatch(2);
-  private final Duration shutdownTimeout = Duration.ofMinutes(1);
+  private final Duration shutdownTimeout = Duration.ofSeconds(15);
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java:149]: what's the impact of this change?

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/permissions/PeerPermissionsDenylist.java [hunk 3]
```diff
@@ -55,8 +55,8 @@ public static PeerPermissionsDenylist create(final int maxSize) {
 
   @Override
   public boolean isPermitted(final Peer localNode, final Peer remotePeer, final Action action) {
-    return !EnodeURLImpl.sameListeningEndpoint(localNode.getEnodeURL(), remotePeer.getEnodeURL())
-        && !denylist.contains(remotePeer.getId());
+    return !denylist.contains(remotePeer.getId())
+        && !EnodeURLImpl.sameListeningEndpoint(localNode.getEnodeURL(), remotePeer.getEnodeURL());
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/permissions/PeerPermissionsDenylist.java:59]: what's the impact of this change?

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java [hunk 4]
```diff
@@ -355,13 +375,34 @@ public void onMessage(final Packet packet, final DiscoveryPeer sender) {
...
+        matchInteraction(packet)
+            .ifPresent(
+                interaction -> {
+                  final Optional<ENRResponsePacketData> packetData =
+                      packet.getPacketData(ENRResponsePacketData.class);
+                  final NodeRecord enr = packetData.get().getEnr();
+                  peer.setNodeRecord(enr);
+
+                  Optional<ForkId> maybeForkId = peer.getForkId();
+                  if (maybeForkId.isPresent()) {
```
Comment (macfarla) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java:387]: does this checking only happen if the --Xfilter-on-enr-fork-id is set?

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java [hunk 5]
```diff
@@ -1469,6 +1480,175 @@ public long read() {
...
+    final List<NodeKey> nodeKeys = PeerDiscoveryTestHelper.generateNodeKeys(1);
+    final List<DiscoveryPeer> peers = helper.createDiscoveryPeers(nodeKeys);
+    final ForkIdManager forkIdManager = mock(ForkIdManager.class);
+    DiscoveryPeer sender = peers.get(0);
+    final Packet enrPacket = prepareForForkIdCheck(forkIdManager, nodeKeys, sender, true);
+
+    when(forkIdManager.peerCheck(any(ForkId.class))).thenReturn(true);
+    controller.onMessage(enrPacket, sender);
+
+    Optional<DiscoveryPeer> maybePeer =
```
Comment (mark-terry) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1494]: final

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java [hunk 6]
```diff
@@ -1469,6 +1480,175 @@ public long read() {
...
+    final List<NodeKey> nodeKeys = PeerDiscoveryTestHelper.generateNodeKeys(1);
+    final List<DiscoveryPeer> peers = helper.createDiscoveryPeers(nodeKeys);
+    final ForkIdManager forkIdManager = mock(ForkIdManager.class);
+    DiscoveryPeer sender = peers.get(0);
+    final Packet enrPacket = prepareForForkIdCheck(forkIdManager, nodeKeys, sender, true);
+
+    when(forkIdManager.peerCheck(any(ForkId.class))).thenReturn(false);
+    controller.onMessage(enrPacket, sender);
+
+    Optional<DiscoveryPeer> maybePeer =
```
Comment (mark-terry) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1516]: final

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java [hunk 7]
```diff
@@ -1469,6 +1480,175 @@ public long read() {
...
+  public void shouldStillCallNotifyPeerBondedIfNoForkIdSent() {
+    final List<NodeKey> nodeKeys = PeerDiscoveryTestHelper.generateNodeKeys(1);
+    final List<DiscoveryPeer> peers = helper.createDiscoveryPeers(nodeKeys);
+    DiscoveryPeer sender = peers.get(0);
+    final Packet enrPacket =
+        prepareForForkIdCheck(mock(ForkIdManager.class), nodeKeys, sender, false);
+
+    controller.onMessage(enrPacket, sender);
+
+    Optional<DiscoveryPeer> maybePeer =
```
Comment (mark-terry) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1537]: final

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java [hunk 8]
```diff
@@ -1469,6 +1480,175 @@ public long read() {
...
+
+    final PongPacketData pongRequestPacketData =
+        PongPacketData.create(localPeer.getEndpoint(), mockPacket.getHash(), UInt64.ONE);
+
+    final Packet pongPacket =
+        Packet.create(PacketType.PONG, pongRequestPacketData, nodeKeys.get(0));
+
+    controller.onMessage(pongPacket, sender);
+
+    NodeRecord nodeRecord = createNodeRecord(nodeKeys.get(0), sendForkId);
```
Comment (mark-terry) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1598]: final

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java [hunk 9]
```diff
@@ -1469,6 +1480,175 @@ public long read() {
...
+
+    ENRResponsePacketData enrResponsePacketData =
```
Comment (mark-terry) [inline ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1600]: final

File: besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java [hunk 10]
```diff
@@ -42,163 +42,177 @@
...
-              new ForkId(Bytes.ofUnsignedInt(0x9007bfccL), 11700000L),
-              new ForkId(Bytes.ofUnsignedInt(0xdb63a1caL), 13189133),
-              new ForkId(Bytes.ofUnsignedInt(0x0f6bf187L), 14525000L),
-              new ForkId(Bytes.ofUnsignedInt(0x7fd1bb25L), 0L),
-              new ForkId(Bytes.ofUnsignedInt(0x7fd1bb25L), 0L))
-        });
+  public static class NotParameterized {
+    @Test
+    public void testFromRaw() {
+      ForkId forkId = new ForkId(Bytes.ofUnsignedInt(0xfe3366e7L), 1735371L);
```
Comment (mark-terry) [inline besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java:55]: final

File: besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java [hunk 11]
```diff
@@ -42,163 +42,177 @@
...
+      List<List<Bytes>> forkIdAsBytesList = List.of(forkId.getForkIdAsBytesList());
```
Comment (mark-terry) [inline besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java:56]: final

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java [hunk 12]
```diff
@@ -705,27 +798,27 @@ void updateFilter(final Predicate<Packet> filter) {
...
-     *     executed.
-     */
-    void execute(final long lastTimeout, final int retryCount) {
+    /** Executes the action associated with this state. Sets a "boomerang" timer to itself. */
+    void execute() {
       action.accept(this);
-      if (retryable && retryCount < MAX_RETRIES) {
-        final long newTimeout = retryDelayFunction.apply(lastTimeout);
+      if (retryCount < MAX_RETRIES) {
+        delay = retryDelayFunction.apply(delay);
```
Comment (mark-terry) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java:805]: If `delay` is set to `0`, does this just default to 2000ms delay every time?

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/EIP2124Test.java [hunk 13]
```diff
@@ -25,6 +25,8 @@
 import org.hyperledger.besu.ethereum.eth.ForkIdTestUtil.GenesisHash;
 import org.hyperledger.besu.ethereum.eth.ForkIdTestUtil.Network;
 import org.hyperledger.besu.ethereum.eth.ForkIdTestUtil.PeerCheckCase;
+import org.hyperledger.besu.ethereum.forkid.ForkId;
+import org.hyperledger.besu.ethereum.forkid.ForkIdManager;
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/manager/EIP2124Test.java:29]: Yes, see the end of the file

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java (hunk 2 above)
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java:149]: There is a timeout happening if the network cannot shutdown within that time range. If we have to wait more than 15 seconds for this to shutdown we might as well kill it.

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/permissions/PeerPermissionsDenylist.java (hunk 3 above)
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/permissions/PeerPermissionsDenylist.java:59]: Nothing. I will reverse it!

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java (hunk 4 above)
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java:387]: Yes, if we do not want to filter on the forkid we do not send an ENR request, which means that the peer does not send the ENR response.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java [hunk 14]
```diff
@@ -112,11 +112,8 @@ public void registerConnection(
             maxMessageSize,
             clock,
             permissioningProviders);
-    final EthPeer ethPeer = connections.putIfAbsent(peerConnection, peer);
-    LOG.debug(
-        "Adding new EthPeer {} {}",
-        peer.getShortNodeId(),
-        ethPeer == null ? "for the first time" : "");
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java:119]: This is replaced with "LOG.debug("Adding new EthPeer {}", peer.nodeId());" in line 116 of EthPeers

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java (hunk 2 above)
Comment (macfarla) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/network/DefaultP2PNetwork.java:149]: 👍

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/DiscoveryPeer.java [hunk 15]
```diff
@@ -136,6 +139,21 @@ public Optional<NodeRecord> getNodeRecord() {
 
   public void setNodeRecord(final NodeRecord nodeRecord) {
     this.nodeRecord = nodeRecord;
+
+    Object rawForkId = nodeRecord.get("eth");
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/DiscoveryPeer.java:143]: done

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/DiscoveryPeer.java [hunk 16]
```diff
@@ -136,6 +139,21 @@ public Optional<NodeRecord> getNodeRecord() {
...
+    if (rawForkId != null) {
+      try {
+        @SuppressWarnings("unchecked")
+        final List<List<Bytes>> typedRawForkId = (List<List<Bytes>>) rawForkId;
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/DiscoveryPeer.java:147]: done

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkId.java [hunk 17]
```diff
@@ -99,6 +108,8 @@ public boolean equals(final Object obj) {
 
   @Override
   public int hashCode() {
-    return super.hashCode();
+    int hashCode = 31;
```
[the remark this answers is not in the document]
Comment (pinges) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/forkid/ForkId.java:111]: This used to use the hashCode method on Object, but as we are using it as the key for hash maps I had to change it to something meaningful. Simplified this method.

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java (hunk 5 above)
Comment (pinges) [reply ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1494]: done

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java (hunk 6 above)
Comment (pinges) [reply ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1516]: done

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java (hunk 7 above)
Comment (pinges) [reply ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1537]: done

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java (hunk 8 above)
Comment (pinges) [reply ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1598]: done

File: ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java (hunk 9 above)
Comment (pinges) [reply ethereum/p2p/src/test/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryControllerTest.java:1600]: done

File: besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java (hunk 10 above)
Comment (pinges) [reply besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java:55]: done

File: besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java (hunk 11 above)
Comment (pinges) [reply besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java:56]: done

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java (hunk 12 above)
Comment (pinges) [reply ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java:805]: delay is a field (I changed the code here to this.delay to make that clearer) and as such the delay will be updated every time the execute is called.

File: besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java (hunk 1 above)
Comment (pinges) [reply besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java:1180]: Setting this changes the behaviour to always sending the last ForkId based on the genesis file and not the ForkId of where we up to at the moment.

**Review by macfarla — APPROVED** [review]

File: besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java (hunk 1 above)
Comment (fab-10) [reply besu/src/main/java/org/hyperledger/besu/RunnerBuilder.java:1180]: for what is it used?

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/NetworkingOptions.java [hunk 18]
```diff
@@ -68,6 +69,13 @@ public class NetworkingOptions implements CLIOptions<NetworkingConfiguration> {
       description = "Whether to enable P2P Discovery Protocol v5 (default: ${DEFAULT-VALUE})")
   private final Boolean isPeerDiscoveryV5Enabled = false;
 
+  @CommandLine.Option(
+      names = FILTER_ON_ENR_FORK_ID,
+      hidden = true,
+      defaultValue = "false",
+      description = "Whether to enable filtering of peers based on the ENR field ForkId)")
+  private final Boolean filterOnEnrForkId = false;
+
```
Comment (fab-10) [inline besu/src/main/java/org/hyperledger/besu/cli/options/unstable/NetworkingOptions.java:78]: is there any risk enabling it by default?

File: ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java [hunk 19]
```diff
@@ -558,11 +646,15 @@ private void findNodes(final DiscoveryPeer peer, final Bytes target) {
...
+    final Bytes id = peer.getId();
+    final PeerInteractionState previous =
+        inflightInteractions
+            .computeIfAbsent(id, k -> new ConcurrentHashMap<>())
+            .put(state.expectedType, state);
     if (previous != null) {
       previous.cancelTimers();
     }
-    state.execute(0, 0);
+    state.execute();
```
Comment (fab-10) [inline ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java:657]: why removed the reset of the counters? could not happen that a already used state is reused?

**Review by fab-10 — COMMENTED** [review]
LGTM, just a couple of questions

**Comment by fab-10** [comment]
And this change deserve a nice CHANGELOG entry, that also explain how to enabled it :)

**Pull request merged**